Add 5, 6 & 7 item shopsanity.

Pepe20129 opened the pull request:
Doesn't add full 8 item shopsanity due to logic assumptions about the renewability of shields.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2006133500.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2006232627.zip)
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2006233442.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2006233606.zip)
<!--- section:artifacts:end -->

Pepe20129 committed:
Format shops.cpp

Pepe20129 committed:
Add 5, 6 & 7 item shopsanity

Pepe20129 committed:
Guarantee a bombchu refill

Dreamwrecker commented:
On 3 separate occasions (kokiri, kak, and goron shop) I bought something, left, came back and the shop would be entirely sold out except for the replenishing item.  

![Shopsanity 567 All sold out adult kak](https://github.com/user-attachments/assets/be7a7a14-bcac-4911-8192-36eda2f8898e)

![Shopsanity 567 All sold out](https://github.com/user-attachments/assets/6feb6e11-f59f-45ab-bb3c-8139393814e2)

I'm not 100% sure, but I think it was specifically the 7th item causing the issue (right side, bottom left).

Pepe20129 committed:
Fish first shop item index

Pepe20129 committed:
Clean up NonShopItems

Pepe20129 commented:
> On 3 separate occasions (kokiri, kak, and goron shop) I bought something, left, came back and the shop would be entirely sold out except for the replenishing item.
> 
> ![Shopsanity 567 All sold out adult kak](https://private-user-images.githubusercontent.com/176975393/356867806-be7a7a14-bcac-4911-8192-36eda2f8898e.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MjI2NDAsIm5iZiI6MTcyMzgyMjM0MCwicGF0aCI6Ii8xNzY5NzUzOTMvMzU2ODY3ODA2LWJlN2E3YTE0LWJjYWMtNDkxMS04MTkyLTM2ZWRhMmY4ODk4ZS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxNlQxNTMyMjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNjY2M2YyYmI0ZjEyMjFjNGVhYjdhZTZiOGMyMDJmNDgxNjA2MDM2NzExNWMwZTVkYWEzMDQ3OTU4YmQ5NjRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.IKScxqwgOQDUPYpTMyHes_mPJula7tMIcs2aLcXTYS0)
> 
> ![Shopsanity 567 All sold out](https://private-user-images.githubusercontent.com/176975393/356867819-6feb6e11-f59f-45ab-bb3c-8139393814e2.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MjI2NDAsIm5iZiI6MTcyMzgyMjM0MCwicGF0aCI6Ii8xNzY5NzUzOTMvMzU2ODY3ODE5LTZmZWI2ZTExLWY1OWYtNDVhYi1iYjNjLTgxMzkzOTM4MTRlMi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxNlQxNTMyMjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMDk2Y2ViYzA5NTQxNzk1ZjdlNzEyZTYyMDdmMWYxM2YzN2I2OGRjMGI2YTJjNTc2YTVmMjVkODQ0YTk4ZTJkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Cu5VsZ8_NOtwsFy9mlRkQjNXcYIWTAxax5nlMPTFVA4)
> 
> I'm not 100% sure, but I think it was specifically the 7th item causing the issue (right side, bottom left).

Fixed and yes, it was that item.

Dreamwrecker commented:
Post fix run went well, no issues found.

Pepe20129 committed:
Merge branch 'develop-rando' into full_shopsanity

Pepper0ni commented on soh/soh/Enhancements/randomizer/3drando/shops.cpp:
```diff
@@ -31,82 +31,85 @@ void PlaceVanillaShopItems() {
+        RG_BUY_BOMBCHUS_10,
+        RG_BUY_DEKU_STICK_1,
+        //^First 8 items: Exist on shopsanity 7 or less
+        RG_BUY_DEKU_STICK_1,
```
There should be much more variety in the second section then a bunch of arrows: Having Green potion is the only non-drop, non-granny source of magic refills. Having a Red Potion and Heart for no-drops healing would be nice. Blue fire because buy blue fire is something that can come up in logic. Fairies can matter once fairy shuffle is in.  After all that, a variety of ammo would make shops much more useful in general.

Pepe20129 replied:
Added all except heart since there was none in the 28 items from n64 rando.

Pepper0ni commented on soh/soh/Enhancements/randomizer/settings.cpp:
```diff
@@ -97,7 +97,7 @@ void Settings::CreateOptions() {
     mOptions[RSK_LINKS_POCKET] = Option::U8("Link's Pocket", {"Dungeon Reward", "Advancement", "Anything", "Nothing"}, OptionCategory::Setting, CVAR_RANDOMIZER_SETTING("LinksPocket"), "", WidgetType::Combobox, RO_LINKS_POCKET_DUNGEON_REWARD);
     mOptions[RSK_SHUFFLE_SONGS] = Option::U8("Shuffle Songs", {"Song Locations", "Dungeon Rewards", "Anywhere"}, OptionCategory::Setting, CVAR_RANDOMIZER_SETTING("ShuffleSongs"), mOptionDescriptions[RSK_SHUFFLE_SONGS], WidgetType::Combobox, RO_SONG_SHUFFLE_SONG_LOCATIONS);
-    mOptions[RSK_SHOPSANITY] = Option::U8("Shopsanity", {"Off", "0 Items", "1 Item", "2 Items", "3 Items", "4 Items", "Random"}, OptionCategory::Setting, CVAR_RANDOMIZER_SETTING("Shopsanity"), mOptionDescriptions[RSK_SHOPSANITY], WidgetType::Combobox, RO_SHOPSANITY_OFF);
+    mOptions[RSK_SHOPSANITY] = Option::U8("Shopsanity", {"Off", "0 Items", "1 Item", "2 Items", "3 Items", "4 Items", "5 Items", "6 Items", "7 Items"/*, "8 Items (All)"*/, "Random"}, OptionCategory::Setting, CVAR_RANDOMIZER_SETTING("Shopsanity"), mOptionDescriptions[RSK_SHOPSANITY], WidgetType::Combobox, RO_SHOPSANITY_OFF);
```
Ideally, this would be a fixed number on the dropdown that spawns a drag bar, similar to how dungeon rewards work

Pepe20129 replied:
Done.

Pepper0ni reviewed (changes requested):
Some small things before I approve.

Pepe20129 committed:
Split count options

Pepe20129 committed:
Improve item ordering

Ensure potions, blue fire and fairy on shopsanity 6 and less.
There are no hearts in the first 28 items (the ones from n64 rando).

Pepe20129 committed:
Merge branch 'develop-rando' into full_shopsanity

Pepe20129 committed:
Post-merge fixes
